src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs

```diff
@@ -399,11 +458,11 @@ public DependencyContext Build(string[] userRuntimeAssemblies = null)
             targetInfo,
             _compilationOptions ?? CompilationOptions.Default,
             compilationLibraries,
-            runtimeLibraries,
+            runtimeLibraries.Select(library => library.Library),
```

Comment:
Since the library and its dependencies are actually immutable, I think you need to recreate each library after the graph walk so that any dependencies that were removed don't show up as dangling links in the deps file.